File: Solutions/227_Basic_Calculator_II/227_Basic_Calculator_II.c
//https://leetcode.com/problems/basic-calculator-ii/discuss/418670/Simple-C-solution.
//Second fastest
int calculate(char * s){
    int len=strlen(s)+3;
    int *num=malloc(sizeof(int)*len/2);
    int total=0,idx=0,i=0,j=0;
    char *stack=malloc(len);
    while(s[i])
    {
        switch(s[i])
        {
            case '+': case '-':
                stack[++j]=s[i];
                break;
            case '*': 
                i++;
                while(s[i]==' ')
                    i++;
                while(s[i]>='0'&&s[i]<='9')
                {
                    int n=s[i++]-'0';
                    total=total*10+n;
                }
                num[++idx]=total;
                total=0;
                i--;
                num[idx-1]=num[idx-1]*num[idx];
                idx--;
                break;
            case '/':
                i++;
                while(s[i]==' ')
                    i++;
                while(s[i]>='0'&&s[i]<='9')
                {
                    int n=s[i++]-'0';
                    total=total*10+n;
                }
                num[++idx]=total;
                total=0;
                i--;
                num[idx-1]=num[idx-1]/num[idx];
                idx--;
                break;
            case '0' ... '9':
                while(s[i]>='0'&&s[i]<='9')
                {
                    int n=s[i++]-'0';
                    total=total*10+n;
                }
                num[++idx]=total;
                total=0;
                i--;
                break;
        }
         i++;
                
    }

    int tmp=j;
    j=1;
    idx=1;
    while(j <= tmp) { 
        if(stack[j]=='+')
            num[idx+1]=num[idx]+num[idx+1];
        else
            num[idx+1]=num[idx]-num[idx+1];
        idx++;j++;
    }
    return num[idx];

}

//Another
int calculate(char * s){
    char sign = '+';
    int i, lastNum = 0, result = 0;
    long curNum = 0;
    int size = strlen(s);
    
    if (size ==  0)
        return 0;
    
    for (i = 0; i < size; i++) {
        char curChar = s[i];
        
        if (isdigit(curChar))
            curNum = curNum * 10 + curChar - '0';
        
        if (!isdigit(curChar) && !iswspace(curChar) || i == size - 1) {
            if (sign == '+' || sign == '-') {
                result += lastNum;
                lastNum = (sign == '+') ? curNum : -curNum;
            } else if (sign == '*') {
                lastNum *= curNum;
            } else if (sign == '/') {
                lastNum /= curNum;
            }
            
            sign = curChar;
            curNum = 0;
        }
    }
    result += lastNum;
    
    return result;
}

//Fastest
struct DataNode {
    int val;
    char op;
};

int calculate(char * s){
    struct DataNode stack[2];
    int top = 0;
    int val = 0;
    
    while (*s != '\0') {
        //printf("%d %c top %d tv %d\n", val, *s, top, top > 0 ? stack[top-1].val: 0);
        if (*s == '+' || *s == '-') {
            while (top) {
                if (stack[top-1].op == '*') {
                    val = stack[top-1].val * val;
                } else if (stack[top-1].op == '/') {
                    val = stack[top-1].val / val;
                } else if (stack[top-1].op == '+') {
                    val = stack[top-1].val + val;
                } else if (stack[top-1].op == '-') {
                    val = stack[top-1].val - val;
                }
                top--;
            }
            stack[top].val = val;
            stack[top++].op = *s;
            val = 0;
        } else if (*s == '*' || *s == '/') {
            if (top) {
                if (stack[top-1].op == '*') {
                    stack[top-1].val *= val;
                    stack[top-1].op = *s;
                } else if (stack[top-1].op == '/') {
                    stack[top-1].val /= val;
                    stack[top-1].op = *s;
                } else {
                    stack[top].val = val;
                    stack[top++].op = *s;
                }
            } else {
                stack[top].val = val;
                stack[top++].op = *s;
            }
            val = 0;
        } else if (*s == ' ') {
        } else {
            val = val*10+((*s)-'0');
        }
        s++;
    }
    
    while (top) {
        if (stack[top-1].op == '*') {
            val = stack[top-1].val * val;
        } else if (stack[top-1].op == '/') {
            val = stack[top-1].val / val;
        } else if (stack[top-1].op == '+') {
            val = stack[top-1].val + val;
        } else if (stack[top-1].op == '-') {
            val = stack[top-1].val - val;
        }
        top--;
    }
    
    return val;
}

File: Solutions/227_Basic_Calculator_II/227_Basic_Calculator_II.go
//Second fastest
func calculate(s string) int {
    s = strings.ReplaceAll(s, " ", "")
    if len(s) == 0 {
        return 0
    }
    
    start, end := 0, 1
    prev := 0
    result := 0
    for end <= len(s) {
        if end == len(s) || isOperator(string(s[end])) {
            num, _ := strconv.Atoi(s[start:end])
            if start == 0 {
                result += num
                prev = num
            } else {
                oper := string(s[start-1])
                if oper == "-" {
                    result -= num
                    prev = -num
                } else if oper == "+" {
                    result += num
                    prev = num
                } else if oper == "*" {
                    result = result - prev + prev*num
                    prev = prev*num
                } else if oper == "/" {
                    result = result - prev + prev/num
                    prev = prev/num
                }
            }
            start = end+1
        }
        end++
    }
    
    return result
}

func isOperator(s string) bool {
    if s == "-" || s == "+" || s == "/" || s == "*" {
        return true
    }
    return false
}

//https://leetcode.com/problems/basic-calculator-ii/discuss/947927/Easy-to-follow-go-golang-solution-100-time
//Fastest
func calculate(s string) int {

	if len(s) == 0 {
		return 0
	}

	cal := 0
	var sign byte = '+'
	//sign := '+' // when I declare sign this way it is not working. Not sure why
	var stack []int

	for i := 0; i < len(s); i++ {
		if IsDigit(s[i]) {
			cal = cal * 10 + int(s[i]-'0')
		}
		if (!IsDigit(s[i]) && s[i] != ' ') || i == len(s)-1 {
			if sign == '+' {
				stack = append(stack, cal)

			} else if sign == '-' {
				stack = append(stack, -cal)
			} else if sign == '*' {
				stack[len(stack)-1] = stack[len(stack)-1] * cal
			} else if sign == '/' {
				stack[len(stack)-1] = stack[len(stack)-1] / cal
			}
			cal = 0
			sign = s[i]
		}
	}
	sum := 0
	for _, v := range stack {
		sum += v
	}
	return sum
}

func IsDigit(s byte) bool {
	if s >= '0' && s <= '9' {
		return true
	}
	return false
}

//Another
func calculate(s string) int {
    
    nums := make([]int, 0)
    curNum := 0
    prevOp := '+'
    s = s + "+"
    
    for i := 0; i < len(s); i++ {
        if s[i] >= '0' && s[i] <= '9' {
            curNum = curNum * 10 + int(s[i] - '0')
        } else if s[i] == ' ' {
            continue
        } else {
            if prevOp == '+' {
                nums = append(nums, curNum)
            } else if prevOp == '-' {
                nums = append(nums, -1 * curNum)
            } else if prevOp == '*' {
                nums[len(nums)-1] = nums[len(nums) - 1] * curNum
            } else if prevOp == '/' {
                nums[len(nums)-1]  =nums[len(nums) - 1] / curNum
            }
            prevOp = rune(s[i])
            curNum = 0
        }
    }
    
    sum := 0
    for _, num := range(nums) {
        sum += num
    }
    
    return sum
    
}